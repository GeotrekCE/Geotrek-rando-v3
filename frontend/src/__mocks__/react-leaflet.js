const MapContainer = ({ children }) => <div data-testid="MapContainer">{children}</div>;

const useMap = () => ({ fitBounds: () => {} });

const TileLayer = () => <div data-testid="TileLayer" />;

const Marker = ({ children }) => <div data-testid="Marker">{children}</div>;

const Popup = () => <div data-testid="Popup" />;

export { MapContainer, TileLayer, Marker, Popup, useMap };
