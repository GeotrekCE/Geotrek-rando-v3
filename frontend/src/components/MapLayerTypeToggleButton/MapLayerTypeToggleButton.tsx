import { getMapConfig } from 'components/Map/config';
import { SquaredButtonWithImage } from 'components/SquaredButtonWithImage/SquaredButtonWithImage';
import L, { ControlPosition } from 'leaflet';
import React, { FunctionComponent, useState } from 'react';
import { useMap } from 'react-leaflet';
import Control from 'components/Map/components/CustomControl';

export type TileLayerType = 'classic' | 'satellite';

interface MapLayerTypeToggleButton {
  position?: ControlPosition;
}

const TILE_LAYERS: TileLayerType[] = ['classic', 'satellite'];

export const MapLayerTypeToggleButton: FunctionComponent<MapLayerTypeToggleButton> = ({
  position = 'bottomleft',
}) => {
  const [tileLayerType, setTileLayerType] = useState<TileLayerType>('classic');

  const otherLayer = TILE_LAYERS.find(buttonsType => buttonsType !== tileLayerType);

  const map = useMap();

  const mapConfig = getMapConfig();

  const isSatelliteLayerAvailable =
    mapConfig.mapSatelliteLayerUrl !== undefined && navigator.onLine;

  if (!otherLayer || !isSatelliteLayerAvailable) {
    return null;
  }

  const updateTileLayer = (newTileLayerType: TileLayerType) => {
    if (map !== undefined) {
      map.eachLayer(layer => {
        if (layer instanceof L.TileLayer) {
          if (newTileLayerType === 'classic') {
            layer.setUrl(mapConfig.mapClassicLayerUrl);
          }
          if (mapConfig.mapSatelliteLayerUrl !== undefined && newTileLayerType === 'satellite') {
            layer.setUrl(mapConfig.mapSatelliteLayerUrl);
          }
        }
      });
    }
  };

  return (
    <Control position={position}>
      <div
        className="leaflet-toggleLayer"
        key={otherLayer}
        onClick={() => {
          setTileLayerType(otherLayer);
          updateTileLayer(otherLayer);
        }}
      >
        <SquaredButtonWithImage
          titleKey={`map.layerButton.${otherLayer}`}
          imageUrl={`/images/${otherLayer}-toggle-button-image.png`}
        />
      </div>
    </Control>
  );
};
