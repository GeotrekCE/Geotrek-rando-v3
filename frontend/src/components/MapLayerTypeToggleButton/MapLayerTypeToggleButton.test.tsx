import React from 'react';
import { MapContainer } from 'react-leaflet';
import { render } from 'services/testing/reactTestingLibraryWrapper';
import { MapLayerTypeToggleButton } from './MapLayerTypeToggleButton';

jest.mock('components/Map/config', () => ({
  getMapConfig: () => ({
    mapSatelliteLayerUrl: 'satelliteURL',
  }),
}));

describe('MapLayerTypeToggleButton', () => {
  it('renders correctly', () => {
    const mapLayerTypeToggleButton = render(
      <MapContainer attributionControl={false} zoomControl={false}>
        <MapLayerTypeToggleButton onChange={() => null} />
      </MapContainer>,
    );

    expect(mapLayerTypeToggleButton).toMatchSnapshot();
  });
});
