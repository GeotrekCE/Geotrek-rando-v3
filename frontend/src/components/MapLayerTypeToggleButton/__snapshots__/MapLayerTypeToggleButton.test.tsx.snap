// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLayerTypeToggleButton renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="leaflet-container leaflet-grab leaflet-touch-drag"
        style="position: relative;"
        tabindex="0"
      >
        <div
          class="leaflet-pane leaflet-map-pane"
          style="left: 0px; top: 0px;"
        >
          <div
            class="leaflet-pane leaflet-tile-pane"
          />
          <div
            class="leaflet-pane leaflet-shadow-pane"
          />
          <div
            class="leaflet-pane leaflet-overlay-pane"
          />
          <div
            class="leaflet-pane leaflet-marker-pane"
          />
          <div
            class="leaflet-pane leaflet-tooltip-pane"
          />
          <div
            class="leaflet-pane leaflet-popup-pane"
          />
        </div>
        <div
          class="leaflet-control-container"
        >
          <div
            class="leaflet-top leaflet-left"
          />
          <div
            class="leaflet-top leaflet-right"
          />
          <div
            class="leaflet-bottom leaflet-left"
          >
            <div
              class="leaflet-control"
            >
              <div
                class="leaflet-toggleLayer"
              >
                <div
                  class="h-12 w-12 relative overflow-hidden shadow-md rounded-medium border-2 border-solid border-white cursor-pointer"
                >
                  <img
                    class="object-cover h-full w-full rounded-medium"
                    src="/images/satellite-toggle-button-image.png"
                  />
                  <span
                    class="font-bold text-white text-CTA text-center w-full absolute bottom-1 z-text"
                  >
                    Satellite
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="leaflet-bottom leaflet-right"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="leaflet-container leaflet-grab leaflet-touch-drag"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        />
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        />
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        />
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        >
          <div
            class="leaflet-control"
          >
            <div
              class="leaflet-toggleLayer"
            >
              <div
                class="h-12 w-12 relative overflow-hidden shadow-md rounded-medium border-2 border-solid border-white cursor-pointer"
              >
                <img
                  class="object-cover h-full w-full rounded-medium"
                  src="/images/satellite-toggle-button-image.png"
                />
                <span
                  class="font-bold text-white text-CTA text-center w-full absolute bottom-1 z-text"
                >
                  Satellite
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="leaflet-bottom leaflet-right"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
